### DISCUSSION TITLE
Why is ARGO Workflows implemented as a Kubernetes Custom Resource?

### DISCUSSION THREAD

**Discussion by johnmezzetta** [argoproj/argo-workflows#10982] [Q&A]
This may be a silly question, but its not obvious why the ARGO workflow architects/designers chose to implement using Kubernetes Custom Resources. If you read the Custom Resource documentation, such an implementation is usually geared to be a Declarative mechanism to support an existing Application Deployment (or Daemon set...). However, ARGO Workflows and use a custom resource approach for its application instead of the traditional Kubernetes Deployment methodology. What were the strategic/technical reasons for this?

**Comment by fhochleitner** [comment]
the big benefit is that you can manage your workflows and templates kubernetes natively. for example define them and deploy them via the gitops engine of your choice.

**Reply by johnmezzetta** [reply]
Thanks for the response Felix. Aren't kubernetes deployments also "kubernetes native"? I might be wrong, but I interpret your use of the term "kubernetes native" as being "the Control Plane API". But that cannot be the only reason? A deployed application API (that flows through the application ingress) can just as easily take ARGO Workflow YAML files as input and spin up the pods (containing the Init and Regular containers) to perform the work. Gitops engines (or orchestrators) aren't limited to using kubernetes control plane APIs; they can surf to any web URL to perform work. Its still not clear why ARGO Workflows was implemented using custom resources... 

I am looking for the strategic/technical reasons why this methodology was selected over regular deployments. And the reasons should be formulated  using words such as "<aspect X> is impractical or impossible to implement using the Deployment methodology". So far, I cannot find any compelling reason to implement what looks like a regular application via custom resources. Any regular deployed application can spin up pods using the kubernetes client API; so there is no limitation there. Still looking for the deeper insight why ARGO Workflows selected this methodology...

**Reply by fhochleitner** [reply]
because a workflow instance does not really on a single deployment. you have quite a complex logic and possibilities for sequences or DAG workflow execution. retry strategy. calling other templates from your workflow, sharing parameters from one node to another, etc. etc. etc. this is why the argo workflow controller exists. there's a lot more code involved in running workflow features and executing them, etc. having the possibility to archive your runs, automatically send logs to archive storage in s3 etc. 
and it's "easy" to define this in a custom resource and have the controller figure out what to do with it. having to do this with deployments, is not feasible, nor would it ever be usuable for anyone.

**Reply by johnmezzetta** [reply]
So your are saying that the "argoexec" GO application (that has all the complex logic you refer to) can only execute as a Kubernetes custom resource and not as a regular deployed application? This implies that custom resource controllers have access to special kubernetes client libraries (and associated kubernetes APIs) not available to regular applications? Can you send a link to a web page with this info?

**Reply by fhochleitner** [reply]
No, that's not what I am saying. What I mean is, that there is an Argo Workflow Controller. Which is actually deployed as a Kubernetes Deployment (see here[https://artifacthub.io/packages/helm/argo/argo-workflows?modal=template&template=controller/workflow-controller-deployment.yaml]  for example). This Controller, is watching CustomResources inside the Cluster, e.g. (Cluster)WorkflowTemplates. And Also Workflows. A workflow instantiates either a workflow defined directly inside this resource, or other templates. 
It then configures and creates pods inside of kubernetes, that run your workflow workload inside of standard kubernetes objects (with the main and wait container, as well as the thing you defined for that node). 

And you have all that complexity inside of the "Workflow" and "WorkflowTemplate" and the creation of which pods to create when, what they are supposed to to, how they might share workspaces, have metrics exposed for the single nodes, etc. etc. Why would you ever build something like that with standard deployments or pods?

**Reply by fhochleitner** [reply]
https://www.youtube.com/watch?v=j5i00z3QXyU this might also be helpful for your understanding on why something is implemented with controllers and custom resources

**Reply by johnmezzetta** [reply]
Hi Felix,

Thanks for the info - you obviously know a lot about ARGO workflows; I am getting closer to understanding. 

The video is interesting, but i must admit that i almost stopped watching since it took 15 minutes until he started to get to the point. Some of what he discussed resonated, but some of it was a stretch imho. He talks about the operator pattern (desired state definition versus actual state and associated remediation). The main issue i have is that he is trying to fit everything into that paradigm, just as the OO paradigm tried to make everything objects and the functional paradigm tries to make everything functions. The reality is that there is a place for everything. Generally speaking, paradigms need to be mixed and matched for optimal system solutions, imho. For example, Infrastructure as Code and representing a Kubernetes Application Deployment in a Git repository with remediation (via ArgoCD or FLUX agents) between the desired state (git definition) and the actual state (the Kubernetes deployment) is a fantastic concept. The video presenter mentions this at 21:29, but does not mention what is being "synchronized". This is a this very good example of the operator pattern employed via cloud native technologies. It fits the bill. But the statement that everything needs to be defined via kubernetes custom resources using the operator pattern is a big stretch imho. Even for the very applicable ArgoCD/Flux case, the actual desired state is defined in by the user in git, not in a kubernetes custom resource (although the implementation might use custom resources - haven't looked).

I will repeat my original question with the technical detail we have established.

Is it fair to say that fundamentally a custom resource is simply some defined data (that is a first class citizen in the control plane API); that the data does not do anything aside from simply being data in a well defined object? Usually that data defines a desired state, but is not limited to this definition. Such data typically relies on separate logic (typically referred to as a controller) to act on the data? You have clarified that the separate logic in question is in an application deployment (specifically the argoexec GO program). So my more succinct question is why are workflows (encapsulated by the workflow yaml) communicated to argoexec using custom resources as opposed to using a regular application API? In other words, why doesn't the argoexec controller pod have a HTTP endpoint that can receive the yaml workflows? Isn't utilization of a custom resource data object a roundabout way to communicate data to an application versus a direct application API call? 

Applications take yaml (or json or xml) in HTTP requests all the time. I have not seen many (i.e. any other) applications take yaml input via Kubernetes custom resources. I see the value of custom resources to control things like Application configuration parameters or upgrade methodology - things outside the normal scope of regular application processing. But ARGO YAML workflows submission seems like a regular part of the argo exec application processing? Furthermore, implementation using the operator pattern in the context of workflow execution seems awkward (if that is what is being done). A workflow is inherently a function, not a desired/actual state (or an object with methods).

This page has criteria the guides the developer in choosing if their app should use custom resources: https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#declarative-apis

ARGO Workflows application API clearly does not fit the declarative model as defined on that page. But on the other hand, custom resources do offer unique features outlined here: https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#common-features. 

The reason I am asking this is not to critique the ARGO Workflow implementation - but to understand how a similar architectural approach might be applicable to my own applications.

**Reply by fhochleitner** [reply]
I think I better understand where you are coming from now. 
Let me start by telling you why I like the way argo workflows handles it's resources as CRDs and what I don't like as much. But first of all let me say this: if you are trying to build something yourself - pick the approach that best works for you. if that's CRDs - great. if it's using synchronous service calls via a REST/gRPC/something-API: also amazing. As you said: pick the tools that fit your usecase not your usecase to fit your tools. 

Now I think it's time we split up our observations regarding the use of CRDs in Argo Workflows in two parts: 
1) (Cluster-)Workflowtemplates
This is the definition of reusable workflows that can be called and instantiated. In my opinion this is a great use of CRDs because it gives me the possibility to define Templates usable for multiple application developer teams to use, I can automatically define them in a git repository and have them rolled out on multiple environments and clusters. I can version them and get all the benefits of gitops. Think of them as Pipeline Definitions in Jenkins but with more flexibility and portability. They are extensible, can call other templates, and I can use them to build my own workflow by chaining together different templates to fit my app-devs specific needs without them needing to reinvent the wheel everytime from scratch. 

2) Workflows
This is the concrete instance of a Workflowtemplate. This is the thing doing the work. Think of it as a runnign pipeline in Jenkins. And I have to agree with you, when I first started using workflows it was very weird for me to have to define this running instance as a kubernetes custom resource. why do I want to use a yaml file to trigger argo to do some work? I want a workflowtemplate to be instantiated on certain triggers. May those triggers be manually from the UI or most likely when a certain event in git happens: a PR is created, I commit a code change, whatever. I don't want to write yaml for this. I don't want to push a workflow-yaml to a repository where my GitOps engine points to just to trigger an application build or whatever. If I understand you correctly that's also your point and what your are trying to point out. And looking at argo workflows alone I completely agree with you. The power of this however, comes from using argo events in conjunction with workflows. You get your usual event sources like webhooks for a git repository, messages to a kafka cluster, and whatnot. You can then transform the data you receive and automatically create this workflow resource and have it execute your workflow. And as it might be weird and not the most intuitive to work with, it also helps observing what is going on inside your clusters with the same tools you use for everything else without the need for context or ui switches. there's also the possibility to do it differently though. you could always react to something you want to do by spinning up a plain kubernetes container running the argo-cli and instantiate a workflow using it. which is probably more akin to what you describe you want to do. 

Hope this helps a bit

**Reply by johnmezzetta** [reply]
Thanks for your insights Felix, they are much appreciated. What you say makes a lot of sense. For the record, i think the ARGO Workflow YAML and the basic way of defining pipeline work is brilliant (container images and associated commands) - I have been looking for an open source solution like this for years and only stumbled upon workflows recently. The features supported by the yaml have focused on the key things that are needed and the syntax is quite elegant, imho. And i get how, generally speaking, workflows should be triggered in the way your describe. Workflows is just a piece of the overall system. 

My question is really only a general Kubernetes question about where the industry might be moving with regards to Kubernetes custom resource definitions and associated usage; trying to understand the thought process behind the workflows implementation choice of using custom resources.  Again, thanks for your insights.
